src/rt/sched/threadpool.h

```diff
@@ -330,67 +336,77 @@ namespace verona::rt
     template<typename... Args>
     void run_with_startup(void (*startup)(Args...), Args... args)
     {
-      active_thread_count = thread_count;
-      T* t = first_thread;
       {
         ThreadPoolBuilder builder(thread_count);
 
         Logging::cout() << "Starting all threads" << Logging::endl;
-        do
+        for (size_t i = 0; i < thread_count; i++)
         {
-          builder.add_thread(&T::run, t, startup, args...);
-          t = t->next;
-        } while (t != first_thread);
+          T* t = threads->popFree();
+          if (t == nullptr)
+            abort();
+          t->setCore(core_pool->cores[i]);
+          threads->addActive(t);
+          builder.add_thread(t->core->affinity, &T::run, t, startup, args...);
+        }
       }
       Logging::cout() << "All threads stopped" << Logging::endl;
 
-      assert(t == first_thread);
-      do
+      T* t = threads->popActive();
+      while (t != nullptr)
+      {
+        T* prev = t;
+        t = threads->popActive();
+        delete prev;
+      }
+      t = threads->popFree();
+      while (t != nullptr)
       {
-        T* next = t->next;
-        delete t;
-        t = next;
-      } while (t != first_thread);
+        T* prev = t;
+        t = threads->popFree();
+        delete prev;
+      }
```

Comment:
Perhaps add this as a method on `SchedulerList`? It is easier to understand the correctness in there.